### PR TITLE
Generate with breaking changes

### PR DESCRIPTION
**Pull request by yejia-stripe**
# Changelog

"⚠️" symbol highlights breaking changes.

### Added

- Add `CheckoutSessionSetupIntentDataParams.Metadata`.
- Add Invoice `UpcomingLines` method.
- Add `ShippingCost` and `ShippingDetails` properties to `CheckoutSession` resource.
- Add `CheckoutSessionShippingCostTax` and `CheckoutSessionShippingCost` classes
- Add `IssuingCardCancellationReasonDesignRejected` constant to `IssuingCardCancellationReason`.
- Add `Validate` field to `Customer` resource.
- Add `Validate` field to `PaymentSourceParams`.
- Add `SetupAttemptPaymentMethodDetailsCardThreeDSecureResultExempted` constant in `SetupAttemptPaymentMethodDetailsCardThreeDSecureResult`.
- Add `SKUPackageDimensionsParams` and `SKUPackageDimensions`.
- Add dedicated structs for different payment sources and transfers.
- Add `Subscription.DeleteDiscount` methods.
- Add `SubscriptionItemUsageRecordSummariesParams`
- Add `UsageRecordSummary` `UsageRecordSummaries`, and `UsageRecordSummaryList` methods in `SubscriptionItem`
- Add `SubscriptionSchedulePhaseBillingCycleAnchor`, `SubscriptionSchedulePhaseBillingCycleAnchorAutomatic`, and `SubscriptionSchedulePhaseBillingCycleAnchorPhaseStart`
- Add `SubscriptionSchedulePhaseInvoiceSettings` and `SubscriptionSchedulePhaseInvoiceSettingsParams `
- `TerminalLocation` `UnmarshalJSON` - make `TerminalLocation` expandable

### ⚠️ Changed

- Rename files to be consistent with the library's naming conventions.
    - `fee.go` to `applicationfee.go` 
    - `fee/client.go` to `applicationfee/client.go` 
    - `sub.go` to `subscription.go` 
    - `sub/client.go` to `subscription/client.go` 
    - `subitem.go` to `subscriptionitem.go` 
    - `subitem/client.go` to `subscriptionitem/client.go` 
    - `subschedule.go` to `subscriptionschedule.go` 
    - `subschedule/client.go` to `subscriptionschedule/client.go` 
    - `reversal.go` to `transferreversal.go` 
    - `reversal/client.go` to `transferreversal/client.go` 

- Change resource names on `client#API` to be plural to be consistent with the library's naming conventions: 
- Rename structs, fields, enums, and methods to be consistent with the library's naming conventions and with the other Stripe SDKs.
  - `Ach` to `ACH`
  - `Acss` to `ACSS`
  - `Bic` to `BIC`
  - `Eps` to `EPS`
  - `FEDEX` to `FedEx`
  - `Iban` to `IBAN`
  - `Ideal` to `IDEAL`
  - `Sepa` to `SEPA`
  - `Wechat` to `WeChat`
  - `ExternalAccount` to `AccountExternalAccount`
  - `InvoiceLine` to `InvoiceLineItem`
  - `Person` structs/enums to use `Person` prefix
  - and others (see Migration guide)

- Change types of various fields in `Account`, `ApplicationFee`, `BalanceTransaction`, `BillingPortalConfiguration`, `Card`, `Charge`, `Customer`, `Discount`, `Invoice`, `Issuing Card`,  `Issuing Dispute `, `Mandate `, `PaymentIntent`, `PaymentMethod`, `Payout`, `Plan `, `Plan `, `Refund`, `SetupIntent`, `Source`, `Source`, `Subscription`, `SubscriptionItem`, `SubscriptionSchedule`, `Terminal ConnectionToken`, `Terminal Location`, `Terminal Reader `, `Topup`, and `Transfer` (see Migration guide).

- Move `BalanceTransaction` iterator from `balance.go` to `balancetransaction.go`
- Fix `BalanceTransactionSource` `UnmarshalJSON` for when `BalanceTransactionSource.Type == "transfer_reversal"` (previously, we were checking if `Type == "reversal"`, which was always false)
- For BankAccount and Card client methods, check that exactly one of `params.Account` and `params.Customer` is set (previously they could both be set, but only one would be used, and it was different between BankAccount and Card)
- Replace `CardVerification` with field-specific enums (with the same values)
- Move `Del` from `discount/client.go` to `customer/client.go` and rename to `DeleteDiscount`
- Move `DelSub` from `discount/client.go` to `subscription/client.go` and rename to `DeleteDiscount`
- Add separate parameter struct for CreditNote `ListPreviewLines` (renamed to `PreviewLines`) method (`[CreditNoteLineItemListPreviewParams -> CreditNotePreviewParams].Lines` `CreditNoteLineParams` -> `CreditNotePreviewLineParams`)
- Replace `FeeRefundParams.ApplicationFee` with `FeeRefundParams.Fee` and `FeeRefundParams.ID`
- Add separate parameter struct for Invoice `GetNext` (renamed to `Upcoming`) method (`InvoiceUpcomingParams`, and nested params `InvoiceUpcomingLinesInvoiceItemPriceDataParams`, `InvoiceUpcomingLinesInvoiceItemDiscountParams`, `InvoiceUpcomingLinesDiscountParams`, `InvoiceUpcomingLinesInvoiceItemPeriodParams`). `Upcoming`-only fields `Coupon`, `CustomerDetails`, `InvoiceItems`, `Subscription`, `SubscriptionBillingCycleAnchor`, `Schedule`, `SubscriptionBillingCycleAnchor`, `SubscriptionBillingCycleAnchorNow`, `SubscriptionBillingCycleAnchorUnchanged`, `SubscriptionCancelAt`, `SubscriptionCancelAtPeriodEnd`, `SubscriptionCancelNow`, `SubscriptionDefaultTaxRates`, `SubscriptionItems`, `SubscriptionProrationBehavior`, `SubscriptionProrationDate`, `SubscriptionStartDate`, `SubscriptionTrialEnd`, `SubscriptionTrialEndNow`, and `SubscriptionTrialFromPlan` are removed from `InvoiceParams`.
- Add separate structs for `BillingDetails` and `BillingDetailsParams`: `PaymentMethodBillingDetails`, `PaymentMethodBillingDetailsParams`
- Add separate structs for `PaymentMethodCardNetwork`: `PaymentMethodCardNetworksAvailable`, `PaymentMethodCardNetworksPreferred`

### Deprecated

- The `SKU` resource has been deprecated. This will be replaced by https://stripe.com/docs/api/orders_v2.

### ⚠️ Removed

- Remove `AccountCapability` enum definition. This was not referenced in the library.
- Remove `UnmarshalJSON` for resources that are not expandable: `BillingPortalSession`, `Capability`, `CheckoutSession`, `FileLink`, `InvoiceItem`, `LineItem`, `Person`, `WebhookEndpoint`
- Remove `AccountRejectReason` (was only referenced in `account/client_test.go`, actual `AccountRejectParams.Reason` is `*string`)
- Remove `AccountParams.RequestedCapabilities` (use Capabilities instead: https://stripe.com/docs/connect/account-capabilities)
- Remove `AccountSettingsParams.Dashboard` and `AccountSettingsDashboardParams` (Note: `Dashboard` are still available on `AccountSettings`, but it's not available as parameters for any of the methods)
- Remove `AccountCompany.RegistrationNumber` (Note: `RegistrationNumber` is still available on `AccountCompanyParams`, but is not returned in the response)
- Remove `BalanceTransactionStatus`. It was meant to be an enum, but none of the enum values were defined, so it was just an alias for string.
- Remove `CardParams.AccountType`. `AccountType` does not exist on any client method for Card. It does on BankAccount, which is similar.
- Remove `id` param from CheckoutSessions `ListLineItems`. Use `CheckoutSessionListLineItemsParams.Session` instead.
- Remove `CheckoutSessionLineItemPriceDataRecurringParams.AggregateUsage`, `CheckoutSessionLineItemPriceDataRecurringParams.TrialPeriodDays`, and `CheckoutSessionLineItemPriceDataRecurringParams.UsageType`
- Remove `CheckoutSessionPaymentIntentDataParams.Params`, `CheckoutSessionSetupIntentDataParams.Params`, `CheckoutSessionSubscriptionDataParams.Params`. `Params` should only be embedded in root method struct, and has extraneous fields not applicable to child/sub structs.
- Remove `CheckoutSessionTotalDetailsBreakdownTax.TaxRate`. Use `CheckoutSessionTotalDetailsBreakdownTax.Rate`
- Remove `CheckoutSessionTotalDetailsBreakdownTax.Deleted`
- Remove `CustomerParams.Token`
- Remove `Discount` `APIResource` embed
- Remove `DiscountParams`
- Remove `FilePurposeFoundersStockDocument` (`"founders_stock_document"` option for `File.Purpose`)
- Remove `InvoiceParams.Paid`. Use `invoice.status` to check for status. `invoice.status` is a read-only field.
- Remove `InvoiceParams.SubscriptionPlan` and `InvoiceParams.SubscriptionQuantity` (note: these would have been on `InvoiceUpcomingParams`)
- Remove `InvoiceListLinesParams.Customer` and `InvoiceListLinesParams.Subscription` (these are not available for Invoice `ListLines`, but are available for `List`)
- Remove `IssuingAuthorizationRequestHistoryViolatedAuthorizationControlEntity` and `IssuingAuthorizationRequestHistoryViolatedAuthorizationControlName` (unused enums)
- Remove `IssuingCardSpendingControlsParams.SpendingLimitsCurrency`. `issuing_card` has `currency`, and `issuing_card.spending_controls.spending_limits.amount` will use that currency
- Remove `IssuingDisputeEvidenceServiceNotAsDescribed.ProductDescription`, `IssuingDisputeEvidenceServiceNotAsDescribed.ProductType`, `IssuingDisputeEvidenceServiceNotAsDescribedParams.ProductDescription`, `IssuingDisputeEvidenceServiceNotAsDescribedParams.ProductType`, and `IssuingDisputeEvidenceServiceNotAsDescribedProductType`. `issuing_dispute.evidence.service_not_as_described` does not have `product_description` or `product_type`. `issuing_dispute.evidence.canceled` does.
- Remove `LineItemTax.TaxRate`. Use `LineItemTax.Rate` instead.
- Remove `LineItem.Deleted`
- Remove `LoginLink.RedirectURL`
- Remove `PaymentIntentOffSession` (unused enum)
- Remove `PaymentIntentConfirmParams.PaymentMethodTypes`
- Remove `PaymentMethodFPX.TransactionID`
- Remove `Payout.BankAccount` and `Payout.Card` (These fields were never populated, use `PayoutDestination.BankAccount` and `PayoutDestination.Card` instead)
- Remove `PlanParams.ProductID`. Use `PlanParams.Product.ID` instead.
- Remove `Shipping` and `ShippingRate` properties from `CheckoutSession` resource. Please use `ShippingCost` and `ShippingDetails` properties instead.
- Remove `DefaultCurrency` property from `Customer` resource. Please use `Currency` property instead.
- Remove `Updated` and `UpdatedBy` from `RadarValueList`
- Remove `Name` from `RadarValueListItem`
- Remove `ReviewReasonType` type from `Review` resource. Use `ReviewReason` instead
- Remove `SetupIntentCancellationReasonFailedInvoice` and `SetupIntentCancellationReasonFraudulent` values from `SetupIntentCancellationReason`
- Remove `SigmaScheduledQueryRun.Query`. The field was invalid
- Remove `SKUParams.Description` and `SKU.Description`
- Remove `SourceMandateAcceptanceStatus`, `SourceMandateAcceptanceStatusAccepted`, `SourceMandateAcceptanceStatusRefused`, `SourceMandateNotificationMethod`, `SourceMandateNotificationMethodEmail`, `SourceMandateNotificationMethodManual`, and `SourceMandateNotificationMethodNone`
- Remove `Source.TypeData` and SourceParams and replace with payment method-specific fields (AUBECSDebit, Bancontact, Card, CardPresent, EPS, Giropay, IDEAL, Klarna, Multibanco, P24, SEPACreditTransfer, SEPADebit, Sofort, ThreeDSecure, Wechat) and `Source.AppendTo` method
- Remove `SourceTransaction.CustomerData`. The field was deprecated
- Remove `SourceTransaction.TypeData` and `SourceTransaction.UnmarshalJSON`. Use payment specific fields - Remove `ACHCreditTransfer`, `CHFCreditTransfer`, `GBPCreditTransfer`, `PaperCheck`, and `SEPACreditTransfer`
- Remove `SubscriptionPaymentBehavior`, `SubscriptionPaymentBehaviorAllowIncomplete`, `SubscriptionPaymentBehaviorErrorIfIncomplete`, and `SubscriptionPaymentBehaviorPendingIfIncomplete`
- Remove `SubscriptionProrationBehavior`, `SubscriptionProrationBehaviorAlwaysInvoice`, `SubscriptionProrationBehaviorCreateProrations`, and `SubscriptionProrationBehaviorNone`
- Remove `SubscriptionStatusAll`
- Remove `SubscriptionParams.Card`, `SubscriptionParams.Plan`, and `SubscriptionParams.Quantity`
- Remove `Subscription.Plan` and `Subscription.Quantity`
- Remove `SubscriptionItemParams.ID`. The field was deprecated
- Remove `SubscriptionSchedulePhaseAddInvoiceItemPriceDataRecurringParams` and `SubscriptionSchedulePhaseAddInvoiceItemPriceDataParams`
- Remove `Del` method on `TaxRate`
- Remove `TerminalReaderGetParams`. Use `TerminalReaderParams` instead.
- Remove `TerminalReaderList.Location` and `TerminalReaderList.Status` (Not available for the list, but is available for individual `TerminalReader`s in `TerminalReaderList.Data`)
- Remove `Token.Email` and `TokenParams.Email`
- Remove `TopupParams.SetSource`
- Remove `WebhookEndpointListParams.Created` and `WebhookEndpointListParams.CreatedRange` (use `StartingAfter` from `ListParams`)
- Remove `WebhookEndpoint.Connected`

### CODE REVIEW

**Comment by yejia-stripe** [comment]
r? @dcr-stripe @kamil-stripe @pakrym-stripe 

Aside from the generated code, also updated the tests (mostly renaming things that were renamed in codegen, and removing some unmarshalling and appendto tests which are no longer applicable).

**Comment by dcr-stripe** [comment]
I'll defer this review to @kamil-stripe .

File: radar_earlyfraudwarning.go [hunk 1]
```diff
@@ -64,6 +64,5 @@ type RadarEarlyFraudWarning struct {
 type RadarEarlyFraudWarningList struct {
 	APIResource
 	ListMeta
-	// TODO: rename `Values` to `Data` in a future major version for consistency with other List structs
-	Values []*RadarEarlyFraudWarning `json:"data"`
+	Data []*RadarEarlyFraudWarning `json:"data"`
```
Comment (kamil-stripe) [inline radar_earlyfraudwarning.go:67]: TODO: Changelog

File: transferreversal.go [hunk 2]
```diff
@@ -13,9 +13,9 @@ import "encoding/json"
 // When reversing transfers, you can optionally reverse part of the transfer. You can do so as many times as you wish until the entire transfer has been reversed.
 //
 // Once entirely reversed, a transfer can't be reversed again. This method will return an error when called on an already-reversed transfer, or when trying to reverse more money than is left on a transfer.
-type ReversalParams struct {
-	Params   `form:"*"`
-	Transfer *string `form:"-"` // Included in URL
```
Comment (kamil-stripe) [inline transferreversal.go:18]: Todo

File: setupintent.go [hunk 3]
```diff
@@ -95,6 +93,26 @@ const (
 	SetupIntentPaymentMethodOptionsACSSDebitVerificationMethodMicrodeposits SetupIntentPaymentMethodOptionsACSSDebitVerificationMethod = "microdeposits"
 )
 
+// Frequency interval of each recurring payment.
+type SetupIntentPaymentMethodOptionsBLIKMandateOptionsOffSessionInterval string
```
Comment (kamil-stripe) [inline setupintent.go:97]: Is BLIK part of the next major or this branch could use refreshing? Note: didn't add it to the changelog.

Comment (yejia-stripe) [reply setupintent.go]: This branch and `sdk-release/next-major` were just behind `master`. Updated!

File: WIPv73Changelog.md [hunk 4]
```diff
@@ -0,0 +1,799 @@
...
+
+  - `ExternalAccount` to `AccountExternalAccount`
+    | Resource | Old name | New name |
+    | --- | --- | --- |
+    | Account | `ExternalAccount` | `AccountExternalAccount` |
+    | Account | `ExternalAccountType` | `AccountExternalAccountType` |
+    | Account | `ExternalAccountTypeBankAccount` | `AccountExternalAccountTypeBankAccount` |
+    | Account | `ExternalAccountTypeCard` | `AccountExternalAccountTypeCard` |
+    | Account | `AccountExternalAccountParams` | `AccountExternalAccountParams` |
+    | Account | `ExternalAccountList` | `AccountExternalAccountList` | <!-- ---- R, S -->
```
Comment (dcr-stripe) [inline WIPv73Changelog.md:370]: What are the `<!-- ---- R, S -->` ?

File: WIPv73Changelog.md [hunk 5]
```diff
@@ -0,0 +1,799 @@
...
+- Add separate structs for `BillingDetails` and `BillingDetailsParams`: `PaymentMethodBillingDetails`, `PaymentMethodBillingDetailsParams`
+- Add separate structs for `PaymentMethodCardNetwork`: `PaymentMethodCardNetworksAvailable`, `PaymentMethodCardNetworksPreferred`
+
+<!-- ---- up to R -->
+
+<!-- ---- after S -->
```
Comment (dcr-stripe) [inline WIPv73Changelog.md:695]: Same wonder here  re comments

File: WIPv73Changelog.md [hunk 6]
```diff
@@ -0,0 +1,799 @@
...
+
+  // After
+  params := &stripe.PlanParams{
+    Product: &stripe.PlanProductParams{
+      ID:                  stripe.String("plan_id"),
+    },
+  }
+  ```
+- Remove `Shipping` and `ShippingRate` properties from `CheckoutSession` resource. Please use `ShippingCost` and `ShippingDetails` properties instead.
+- Remove `DefaultCurrency` property from `Customer` resource. Please use `Currency` property instead.
```
Comment (dcr-stripe) [inline WIPv73Changelog.md:775]: Should we set up a standard format for how we want to phrase removes (ie. when do we include remove prefix vs when do we not?)

**Review by dcr-stripe — COMMENTED** [review]
This LGTM overall - amazing work Yejia! I left a few comments on the changelog. Once it's settled is the plan to move it into this PR description so it gets picked up by our release infra and delete the WIP file?

**Comment by yejia-stripe** [comment]
> This LGTM overall - amazing work Yejia! I left a few comments on the changelog. Once it's settled is the plan to move it into this PR description so it gets picked up by our release infra and delete the WIP file?

@dcr Yes

File: WIPv73Changelog.md (hunk 6 above)
Comment (yejia-stripe) [reply WIPv73Changelog.md:775]: Yes, will do today

File: WIPv73Changelog.md (hunk 4 above)
Comment (yejia-stripe) [reply WIPv73Changelog.md:370]: This was to show where to merge in Kamil's stuff (for alphabetizing). Will remove

**Review by kamil-stripe — APPROVED** [review]

File: account.go [hunk 7]
```diff
@@ -270,7 +142,8 @@ const (
 type AccountParams struct {
 	Params `form:"*"`
 	// An [account token](https://stripe.com/docs/api#create_account_token), used to securely provide details to the account.
-	AccountToken *string `form:"account_token"`
+	AccountToken      *string                         `form:"account_token"`
+	BACSDebitPayments *AccountBACSDebitPaymentsParams `form:"bacs_debit_payments"`
```
Comment (remi-stripe) [inline account.go:146]: This is incorrect. This should be inside `Settings`

File: account.go [hunk 8]
```diff
@@ -856,14 +728,12 @@ type AccountSettingsTreasuryParams struct {
 
 // Options for customizing how the account functions within Stripe.
 type AccountSettingsParams struct {
-	BACSDebitPayments *AccountSettingsBACSDebitPaymentsParams `form:"bacs_debit_payments"`
```
Comment (remi-stripe) [inline account.go:859]: This should not disappear, an override must be wrong

File: account.go [hunk 9]
```diff
@@ -872,6 +742,21 @@ type AccountSettingsParams struct {
...
+	// The Unix timestamp marking when the account representative accepted their service agreement.
+	Date *int64 `form:"date"`
+	// The IP address from which the account representative accepted their service agreement.
+	IP *string `form:"ip"`
+	// The user's service agreement type.
+	ServiceAgreement *string `form:"service_agreement"`
+	// The user agent of the browser from which the account representative accepted their service agreement.
+	UserAgent *string `form:"user_agent"`
+}
+type AccountBACSDebitPaymentsParams struct {
```
Comment (remi-stripe) [inline account.go:756]: This is named wrong, it should be AccountSettingsBACSDebitPaymentsParams

File: checkout/session/client.go [hunk 10]
```diff
@@ -127,7 +129,22 @@ func (c Client) ListLineItems(id string, listParams *stripe.CheckoutSessionListL
 	}
 }
 
-type LineItemIter = lineitem.Iter
+// LineItemIter is an iterator for line items.
+type LineItemIter struct {
```
Comment (remi-stripe) [inline checkout/session/client.go:133]: Still unclear why this moved here.

Comment (pakrym-stripe) [reply checkout/session/client.go:133]: Generating iterators as part of clients that return them is the standard behavior. We had a special case for this one iterator. 

Examples of others:

https://github.com/stripe/stripe-go/blob/13752c6a5e72d07d4ae5d1894763bb36c62dedf9/creditnote/client.go#L186
https://github.com/stripe/stripe-go/blob/b8b156bcee1559adff7d4249f702bfe110077c41/paymentlink/client.go

**Review by kamil-stripe — APPROVED** [review]

**Pull request merged**